### ISSUE TITLE
Enhancement Request: Include Relevant File Names in Query Responses

### ISSUE THREAD

**Issue by llmwesee** [h2oai/h2ogpt#1270]
I have a collection of 100 PDF files, and when I submit a query, the system generates responses with sources. However, I suggest enhancing the generated responses to include the corresponding file name that is most relevant to the query. Currently, the sources provided are often irrelevant and contain information that does not align with the user's query, causing confusion and inefficiency. Including relevant file names will significantly improve clarity and organisation, helping users quickly identify the most pertinent sources.

**Comment by pseudotensor** [comment]
I don't quite know what you mean.

The sources list are already the most relevant from semantic search.  The cutoff can sometimes be issue, depending upon your chosen embedding, which can be confusing.

* You can change cut_distance to be smaller from 1.64 to some new value like 1.5 that will be stricter in what is allowed to be used.
* You can change use_llm_if_no_docs=False to avoid a response if no useful docs.
* You can enable HYDE (say level 1) in order to deal with vague queries if the problem is docs exist for the query but the query is too vague for similarity search.

**Comment by llmwesee** [comment]
I have more than 1000 of scanned pdf files. first i create a UserData by running the command:
`python3 src/make_db.py --user_path=user_path--collection_name=UserData --hf_embedding_model=hkunlp/instructor-large`

Then i migrate those embedding:
`python generate.py --base_model=meta-llama/Llama-2-13b-chat-hf  --score_model=None --user_path=user_path --use_auth_token=True --hf_embedding_model=hkunlp/instructor-large --cut_distance=1.5 --max_seq_len=4096 --migrate_embedding_model=True --langchain_modes="['UserData', 'LLM']"`

And after doing some testing then I ran the final command for both the collections:
``python generate.py --base_model=mistralai/Mistral-7B-Instruct-v0.2 --score_model=None --use_auth_token=True --max_seq_len=8192 --max_new_tokens=4096 --max_max_new_tokens=4096 --concurrency_count=64 --prompt_type=mistral --do_sample=True --langchain_modes="['UserData','UserData1','LLM','MyData']" --use_llm_if_no_docs=False --min_new_tokens=128 --cut_distance=1.5``

Something I notice when created the collection:
1. Only 861 documents are shown in **Doc Counts** for Collection: UserData. Why not all 1000 documents are shown & How do i know embedding are successfully created for the remaining documents out of 1000.
2. In **Document Section** source have only index for 763 files. Where are the remaining index? Please clarify this!
3. For remaining pdf files I created another collection: UserData1. I want to query through both the collections at the same time but the UI allows only one collection at a time. How can i query multiple collection at the same time if the embedding are not created successfully for all the PDF files and I have to create for multiple collection for those files.
4. When asking query then it generate irrelevant response with respect to the query. I noticed for response generation it does not taken the right document context because in the given sources list the corresponding information for the query falls into the low score source link but it generate from another source link

**Comment by pseudotensor** [comment]
Hi,

1. When parsing the PDFs, it should say if there are any exceptions and show the exceptions.  So when running the make_db, keep the output and review.   E.g. if you didn't enable DocTR, then image PDFs will be empty if only use pymupdf.  Just another example, if it was a CSV of too large size, it will be rejected too.

If the log output from make_db is too hard to parse, then one can do it from the UI and upload files and see when something fails, which will appear in the Document Selection "Document Exceptions" text box.

3. Yes, it's a feature request to do multiple collections: https://github.com/h2oai/h2ogpt/issues/841

4. Yes, it's a general RAG issue that retrieval and generation models both have to be good.  Mistral v0.2 is an ok model, but it's not the best.  Mixtral with the RAG prompt in h2oGPT does much better, on level of gpt-4 and Claude.

E.g. in older benchmarks when we included v0.1, it didn't do well:

![image](https://github.com/h2oai/h2ogpt/assets/2249614/fb6cc9d0-c018-45a5-a7cf-85131024d770)

I didn't get around to testing v0.2.

**Comment by llmwesee** [comment]
1. I uploaded the files ticking max quality using UI. And I don't get any "Document Exceptions" in the Document Section text box but still Only 861 out of 1000 documents are shown in Doc Counts for Collection: UserData. What is the reason behind this.
2. I try all the things you suggest for improve retrieval except Mixtral 8x7b or llama 70b but still not seeing any improvement than previous one. is there other things available  to try except cut_distance, HYDE, context length, top_k_docs?

**Comment by llmwesee** [comment]
I would like to propose another scenario for consideration. Let's say we are querying a database of 1000 PDFs or docs. I formulate a prompt, and based on that prompt, the system retrieves some sources and generates a response. However, during the response generation process, it selects 10 top_k_docs chunks. Unfortunately, most of these chunks are irrelevant to the query and receive high scores, such as 0.90, and hold high ranks in the sources list.

The challenge arises when the relevant chunks that align with the query are found in the middle, and most of the generated response is derived from them. However, these relevant chunks do not receive high scores and ranks. Each relevant chunk has a corresponding source, and the source, in turn, has a heading and sub-heading.

To address this, I suggest that we enhance the generated response by including the Source Document Name, heading, or sub-heading as a reference from the most relevant chunks in the chat-bot. This adjustment could significantly improve relevancy, enabling us to place greater trust in the generated response. With this enhancement, it becomes easier to verify the accuracy of the generated responses against the relevant documents.

**Comment by pseudotensor** [comment]
Yes, meta data can help, along with other things like example questions.  Just more expensive.

**Comment by pseudotensor** [comment]
I added ability to add meta data from document ingestion.  It's flexible, but requires original document have that metadata.  E.g. no way for user to add meta data, but you could edit the source document to have it.
